static/js/unread_ops.js

```diff
@@ -14,36 +14,98 @@ import * as notifications from "./notifications";
+    args = {
+        anchor: "first_unread",
+        include_anchor: true,
+        messages_read_till_now: 0,
+        ...args,
+    };
+    const request = {
+        anchor: args.anchor,
+        include_anchor: args.include_anchor,
+        num_before: 0,
+        num_after: NUM_OF_MESSAGES_UPDATED_PER_BATCH,
+        op: "add",
+        flag: "read",
+        // Since there's a database index on is:unread, it's a fast
+        // search query and thus worth including here as an optimization.
+        narrow: JSON.stringify([{operator: "is", operand: "unread", negated: false}]),
+    };
     channel.post({
-        url: "/json/mark_all_as_read",
-        success: () => {
-            // After marking all messages as read, we reload the browser.
-            // This is useful to avoid leaving ourselves deep in the past.
-            // This is also the currently intended behavior in case of partial success,
-            // (response code 200 with result "partially_completed")
-            // where the request times out after marking some messages as read,
-            // so we don't need to distinguish that scenario here.
-            // TODO: The frontend handling of partial success can be improved
-            // by re-running the request in a loop, while showing some status indicator
-            // to the user.
-            reload.initiate({
-                immediate: true,
-                save_pointer: false,
-                save_narrow: true,
-                save_compose: true,
-            });
+        url: "/json/messages/flags/narrow",
+        data: request,
+        success(data) {
+            const messages_read_till_now = args.messages_read_till_now + data.updated_count;
+
+            if (!data.found_newest) {
+                // If we weren't able to make everything as read in a
+                // single API request, then show a loading indicator.
+                ui_report.loading(
+                    $t_html(
+                        {defaultMessage: "Working… {N} messages marked as read so far."},
+                        {N: messages_read_till_now},
+                    ),
+                    $("#request-progress-status-banner"),
+                );
+                if (!loading_indicator_displayed) {
+                    loading.make_indicator(
+                        $("#request-progress-status-banner .loading-indicator"),
+                        {abs_positioned: true},
+                    );
+                    loading_indicator_displayed = true;
+                }
+
+                mark_all_as_read({
+                    include_anchor: false,
+                    anchor: data.last_processed_id,
+                    messages_read_till_now,
+                });
+            } else {
+                if (loading_indicator_displayed) {
+                    // Only show the success message if a progress banner was displayed.
+                    ui_report.loading(
```

Comment:
The comment here was incomplete, so changed it.

Reply:
Thanks!